cloud_blobstore/__init__.py

```diff
@@ -1,4 +1,70 @@
 import typing
+import types
+
+
+class PagedIter(typing.Iterable[str]):
+    """
+    Provide an iterator that will iterate over every object, filtered by prefix and delimiter. Alternately continue
+    iteration with token and key (start_after_key).
+    """
+
+    def get_api_response(self, next_token):
+        """
+        Make blobstore-specific list api request.
+        """
+        raise NotImplementedError()
+
+    def get_listing_from_response(self, resp) -> typing.Iterable[str]:
+        """
+        Retrieve blob key listing from blobstore response.
+        """
+        raise NotImplementedError()
+
+    def get_next_token_from_response(self, resp) -> str:
+        """
+        Retrieve opaque continuation token from blobstore response.
+        """
+        raise NotImplementedError()
+
+    def __iter__(self):
+        """
+        Iterate over the blobs, saving page tokens and blob key start_after_keys as needed in order to continue
+        listing where one left off.
+
+        If start_after_key is not None, iteration will begin on the next key if start_after_key is found on the
+        first page of results. If it is not found on the first page of results, BlobPagingError will be raised.
+        """
+        next_token = self.token
+
+        while True:
+            self.token = next_token
+
+            resp = self.get_api_response(next_token)
+            listing = self.get_listing_from_response(resp)
+
+            if self.start_after_key:
+                while True:
+                    try:
+                        key = next(listing)
```

Comment:
Explicit use of next errors if listing returns a sequence.

Reply:
Nice.